lazy_loader/__init__.py

```diff
@@ -160,38 +166,72 @@ def myfunc():
-    try:
-        return sys.modules[fullname]
-    except KeyError:
-        pass
+    module = sys.modules.get(fullname)
+    have_module = module is not None
+
+    # Most common, short-circuit
+    if have_module and require is None:
+        return module
+
+    spec = None
+    if not have_module:
+        spec = importlib.util.find_spec(fullname)
+        have_module = spec is not None
+
+    if not have_module:
+        not_found_message = f"No module named '{fullname}'"
+    elif require is not None:
+        # Old style lazy loading to avoid polluting sys.modules
+        import packaging.requirements
+
+        req = packaging.requirements.Requirement(require)
+        try:
+            have_module = req.specifier.contains(
+                importlib_metadata.version(req.name),
+                prereleases=True,
+            )
+        except importlib_metadata.PackageNotFoundError as e:
+            raise ValueError(
+                f"Found module '{fullname}' but cannot test requirement '{require}'. "
+                "Requirements must match distribution name, not module name."
+            ) from e
 
-    spec = importlib.util.find_spec(fullname)
-    if spec is None:
+        if not have_module:
+            not_found_message = f"No distribution can be found matching '{require}'"
+
+    if not have_module:
         if error_on_import:
-            raise ModuleNotFoundError(f"No module named '{fullname}'")
-        else:
-            try:
-                parent = inspect.stack()[1]
-                frame_data = {
-                    "spec": fullname,
-                    "filename": parent.filename,
-                    "lineno": parent.lineno,
-                    "function": parent.function,
-                    "code_context": parent.code_context,
-                }
-                return DelayedImportErrorModule(frame_data, "DelayedImportErrorModule")
-            finally:
-                del parent
-
-    module = importlib.util.module_from_spec(spec)
-    sys.modules[fullname] = module
-
-    loader = importlib.util.LazyLoader(spec.loader)
-    loader.exec_module(module)
+            raise ModuleNotFoundError(not_found_message)
+        try:
+            parent = inspect.stack()[1]
+            frame_data = {
+                "filename": parent.filename,
+                "lineno": parent.lineno,
+                "function": parent.function,
+                "code_context": parent.code_context,
+            }
+            return DelayedImportErrorModule(
+                frame_data,
+                "DelayedImportErrorModule",
+                message=not_found_message,
+            )
+        finally:
+            del parent
+
+    if spec is not None:
+        module = importlib.util.module_from_spec(spec)
+        sys.modules[fullname] = module
+
+        loader = importlib.util.LazyLoader(spec.loader)
+        loader.exec_module(module)
 
     return module
 
 
+def have_module(module_like: types.ModuleType) -> bool:
```

Comment:
Is `have_module` used anywhere?

Reply:
Not internally. I thought it would be a nice addition to the API, as I will often do something like:

```Python
# Was: h5py, have_h5py, _ = optional_package('h5py')
h5py = lazy.load('h5py')
have_h5py = lazy.have_module(h5py)

needs_h5py = unittest.skipUnless(have_h5py, 'Test requires h5py')

@needs_h5py
def test_mything():
    ...
```

I included it here because I initially had `load_requirement() -> tuple[module, boolean]`, so this replaced that second return value. Could be a separate PR to keep this clean, or I can just do it downstream, myself. The lazy loading and deferred errors are the important things to me.